pyvisa_py/protocols/hislip.py

```diff
@@ -145,96 +393,165 @@ def __init__(self, ip_addr, open_timeout=None, port=PORT):
         self._async_init = self.async_initialize(session_id=init.session_id)
 
         # get the maximum message size
-        max_msg_size = self.async_maximum_message_size(1 << 24)
-        self.max_payload_size = max_msg_size - HEADER_SIZE
+        max_msg_size = self.async_maximum_message_size(MAX_MSG_SIZE)
+        self._max_payload_size = max_msg_size - HEADER_SIZE
 
         self.timeout = 10
-        self.rmt = 0
-        self.receiver = None
-        self.message_id = 0xFFFFFF00
+        self._rmt = 0
+        self._expected_message_id: Optional[int] = None
+        self._message_id = 0xFFFF_FF00
+        self._last_message_id: Optional[int] = None
 
     # ================ #
     # MEMBER FUNCTIONS #
     # ================ #
 
-    def close(self):
+    def close(self) -> None:
         self._sync.close()
         self._async.close()
 
     @property
-    def timeout(self):
+    def timeout(self) -> float:
         """Timeout value in seconds for both the sync and async sockets"""
         return self._timeout
 
     @timeout.setter
-    def timeout(self, val):
+    def timeout(self, val: float) -> None:
         """Timeout value in seconds for both the sync and async sockets"""
         self._timeout = val
         self._sync.settimeout(self._timeout)
         self._async.settimeout(self._timeout)
 
-    def send(self, data):
-        """Sends the data on the synchronous channel.
+    def send(self, data: bytes) -> int:
+        """Send the data on the synchronous channel.
 
         More than one packet may be necessary in order
-        to not exceed max_payload_size.
+        to not exceed _max_payload_size.
         """
+        # print(f"send({data=})")
         data_view = memoryview(data)
         num_bytes_to_send = len(data)
 
-        # send the data in chunks of self.max_payload_size bytes at a time
+        # send the data in chunks of self._max_payload_size bytes at a time
         while num_bytes_to_send > 0:
-            if num_bytes_to_send <= self.max_payload_size:
+            if num_bytes_to_send <= self._max_payload_size:
                 assert len(data_view) == num_bytes_to_send
-                self.send_data_end_packet(data_view)
+                self._send_data_end_packet(data_view)
                 bytes_sent = num_bytes_to_send
             else:
-                self.send_data_packet(data_view[: self.max_payload_size])
-                bytes_sent = self.max_payload_size
+                self._send_data_packet(data_view[: self._max_payload_size])
+                bytes_sent = self._max_payload_size
 
             data_view = data_view[bytes_sent:]
             num_bytes_to_send -= bytes_sent
 
         return len(data)
 
-    def receive(self, max_len=4096):
+    def receive(self, max_len: int = 4096) -> bytes:
```

Comment:
Does HiSLIP gives any warranty that when asked for 1 bytes it will return a single byte ? I know we have such an issue in VXI-11 where we get full messages and we need to buffer internally.